DigitalLearningSolutions.Data.Migrations/202206281412_AddForeignKeysToSupervisorDelegatesAndCandidateAssessments.cs

```diff
@@ -0,0 +1,63 @@
+﻿namespace DigitalLearningSolutions.Data.Migrations
+{
+    using FluentMigrator;
+
+    [Migration(202206281412)]
+    public class AddForeignKeysToSupervisorDelegatesAndCandidateAssessments : Migration
+    {
+        public override void Up()
+        {
+            Alter.Table("SupervisorDelegates")
+                .AddColumn("DelegateUserID").AsInt32().Nullable()
+                .ForeignKey("Users", "ID");
+
+            Alter.Table("CandidateAssessments")
+                .AddColumn("DelegateUserID").AsInt32().Nullable()
+                .ForeignKey("Users", "ID")
+                .AddColumn("CentreID").AsInt32().Nullable()
+                .ForeignKey("Centres", "CentreID");
+
+            Execute.Script(
+                "DigitalLearningSolutions.Data.Migrations/Scripts/HEEDLS-932-PopulateNewKeysOnSupervisorDelegatesAndCandidateAssessments.sql"
+            );
+
+            Alter.Table("CandidateAssessments")
+                .AlterColumn("DelegateUserID").AsInt32().NotNullable()
+                .AlterColumn("CentreID").AsInt32().NotNullable();
+
+            Delete.ForeignKey().FromTable("SupervisorDelegates").ForeignColumn("CandidateId")
+                .ToTable("Candidates").PrimaryColumn("CandidateID");
+            Alter.Table("SupervisorDelegates").AlterColumn("CandidateId").AsInt32().Nullable();
+
+            Delete.ForeignKey().FromTable("CandidateAssessments").ForeignColumn("CandidateID")
+                .ToTable("Candidates").PrimaryColumn("CandidateID");
+            Alter.Table("CandidateAssessments").AlterColumn("CandidateID").AsInt32().Nullable();
+
+            Rename.Column("CandidateID").OnTable("SupervisorDelegates").To("CandidateID_deprecated");
+            Rename.Column("CandidateID").OnTable("CandidateAssessments").To("CandidateID_deprecated");
+        }
+
+        public override void Down()
+        {
+            Rename.Column("CandidateID_deprecated").OnTable("SupervisorDelegates").To("CandidateID");
+            Rename.Column("CandidateID_deprecated").OnTable("CandidateAssessments").To("CandidateID");
+
+            Alter.Table("SupervisorDelegates").AlterColumn("CandidateID").AsInt32().Nullable()
+                .ForeignKey("DelegateAccounts", "ID");
+            Alter.Table("CandidateAssessments").AlterColumn("CandidateID").AsInt32().Nullable()
+                .ForeignKey("DelegateAccounts", "ID");

```

Comment:
(Not sure if this comment will make any sense) It looks like these columns are not currently nullable and they are made nullable by the Up migration. In general, I think I would expect that after applying the Down migration, the database has the same structure as it had before applying the Up migration.

In the Up migration we populate the new columns and make the deprecated columns nullable. It is possible that we then add new data with the deprecated columns as null. If we were to revert the database to the original structure, the deprecated columns should be non nullable and we would have trouble with those new records where they are null. In theory, we could take those values from the new columns and populate the old columns (kinda like a reversed populating script). It sounds like quite a lot of effort, @SteveJacksonSoft do you think it's worth trying?

Reply:
I think this might not be worth worrying about because we should not be inserting data into deprecated columns via the application, so the solution to "it is possible that we then add new data with the deprecated columns as null" is: Don't Do That

Reply:
I discussed this with Steve while coding, we're willing to assume that the down migration is run immediately after the up migration in case of some problem, and that no new data has been added to the DB. So we don't need to worry about null values. If some data is added and then the down migration needs to be run then it will need to be cleaned up manually.

Reply:
No, that was just my misreading my own code - they should be `NotNullable()`, good catch.

Reply:
Correction, only the CA table column should be not nullable. The SA table column is nullable to begin with.